### PR TITLE
Avoid allocating TLS for the first thread that loads. NFC

### PR DESCRIPTION
**Pull request by sbc100**
This change depend on an LLVM-side change in order to actually
do anything: https://reviews.llvm.org/D126107.  It can land on its
own but won't have any effect since `__tls_base` will always be zero
on startup until that change lands.

The linker will load the TLS data into a static location that the
loading thread can use directly rather than dynamically allocating it.
Secondary threads still require dynamic allocation but I'm hoping to
avoid that too (at least for load-time dynamic libraries and especially
for the main module).

See: #16948

### CODE REVIEW

File: system/lib/pthread/emscripten_tls_init.c [hunk 1]
```diff
@@ -21,15 +23,47 @@ extern void __wasm_init_tls(void *memory);
...
 extern int __dso_handle;
 
+// Create a thread-local wasm global which signal that the current thread is non
+// to use the primary/static TLS region.  Once this gets set it forces that all
+// future calls to emscripten_tls_init to dynamically allocate TLS.
+// This is needed because emscripten re-used module instances for new threads
+// which means stale values of __tls_base from a previous pthread need to be
+// ignored.
+__asm__(".globaltype is_not_primary_tls, i32\n"
+        "is_not_primary_tls:\n");
```
Comment (sbc100) [inline system/lib/pthread/emscripten_tls_init.c:33]: First use of inline asm in emscripten code?   Works pretty well!

File: src/postamble.js [hunk 2]
```diff
@@ -425,7 +425,11 @@ function exit(status, implicit) {
       throw 'unwind';
     } else {
 #if PTHREADS_DEBUG
+#if EXIT_RUNTIME
       err('main thread called exit: keepRuntimeAlive=' + keepRuntimeAlive() + ' (counter=' + runtimeKeepaliveCounter + ')');
+#else
+      err('main thread called exit: keepRuntimeAlive=' + keepRuntimeAlive());
+#endif
```
Comment (tlively) [inline src/postamble.js:432]: Is this change related?

File: system/lib/pthread/emscripten_tls_init.c [hunk 3]
```diff
@@ -21,15 +23,47 @@ extern void __wasm_init_tls(void *memory);
 
 extern int __dso_handle;
 
+// Create a thread-local wasm global which signal that the current thread is non
+// to use the primary/static TLS region.  Once this gets set it forces that all
+// future calls to emscripten_tls_init to dynamically allocate TLS.
+// This is needed because emscripten re-used module instances for new threads
```
Comment (tlively) [inline system/lib/pthread/emscripten_tls_init.c:29]:
```suggestion
// This is needed because emscripten re-used module instances for new threads on the same worker
```

File: system/lib/pthread/emscripten_tls_init.c [hunk 4]
```diff
@@ -21,15 +23,47 @@ extern void __wasm_init_tls(void *memory);
...
+// which means stale values of __tls_base from a previous pthread need to be
+// ignored.
```
Comment (tlively) [inline system/lib/pthread/emscripten_tls_init.c:31]: Can we clear `__tls_base` when killing a thread instead? I guess not because it's only accessible via a builtin that only returns its value, right?

File: tests/test_other.py [hunk 5]
```diff
@@ -11349,7 +11349,7 @@ def test_standard_library_mapping(self):
     self.run_process([EMBUILDER, 'build', 'libc-mt-debug', 'libcompiler_rt-mt', 'libdlmalloc-mt'])
 
     libs = ['-lc', '-lcompiler_rt', '-lmalloc']
-    err = self.run_process([EMCC, test_file('hello_world.c'), '-pthread', '-nostdlib', '-v'] + libs, stderr=PIPE).stderr
+    err = self.run_process([EMCC, test_file('hello_world.c'), '-pthread', '-nodefaultlibs', '-v'] + libs, stderr=PIPE).stderr
```
Comment (tlively) [inline tests/test_other.py:11352]: Why this change?

File: src/postamble.js (hunk 2 above)
Comment (sbc100) [reply src/postamble.js:432]: Sorry, no, this just just came up when I was trying to enable PTHREAD_DEBUG in a build without EXIT_RUNTIME.

Do you think its worth splitting out?

File: system/lib/pthread/emscripten_tls_init.c (hunk 4 above)
Comment (sbc100) [reply system/lib/pthread/emscripten_tls_init.c:31]: I looked into that option.. but I feel like this one is more fool proof and cleaner (for now).

In the long run we are going to have to figure out how to free the TLS data of each side module (currently we leak that), so its possible we could combine the freeing with the clearing.  But for now I like this localized approach to the solving the problem which is fool proof and doesn't require a guaranteed clearing of the global.

File: tests/test_other.py (hunk 5 above)
Comment (sbc100) [reply tests/test_other.py:11352]: Because the emscripten_free_tls function is now defined in `crtbegin.o` .. this change means that C runtime libs are skipped but not core startup objects.

File: src/postamble.js (hunk 2 above)
Comment (tlively) [reply src/postamble.js:432]: No, that's fine. Just making sure I'm not misunderstanding anything.

**Review by tlively — APPROVED** [review]
